### ISSUE TITLE
Crash on bogus input

### ISSUE THREAD

**Issue by laslowh** [flavray/avro-rs#40]
Avro-rs has a tendency to crash the process when given invalid input.  The typical error message that I've seen (I don't have it right now) mentions an out of memory condition.  My guess is it's trying to allocate an array of size -7 or something.  I'd like to start using this crate in a project of mine, but the crashes are an issue.

I've actually created a fuzzer project with [afl.rs](https://rust-fuzz.github.io/book/afl.html), that I can share  if that would be helpful, but at its core it's just:

```
extern crate afl;
extern crate avro_rs;

use avro_rs::{Schema, from_avro_datum};

    static SCHEMA: &'static str = r#"
            {
                "type": "record",
                "name": "test",
                "fields": [
                    {"name": "a", "type": "long", "default": 42},
                    {"name": "b", "type": "string"}
                ]
            }
        "#;


fn main() {
    afl::read_stdio_bytes(|bs| {
	let schema = Schema::parse_str(SCHEMA).unwrap();
        let _ = from_avro_datum(&schema, &mut &bs[..], None);
    });
}
```

And here are the stats (4 crashes) after running for a minute or two:

```
start_time        : 1529615156
last_update       : 1529615246
fuzzer_pid        : 27636
cycles_done       : 32
execs_done        : 178688
execs_per_sec     : 2081.67
paths_total       : 18
paths_favored     : 7
paths_found       : 17
paths_imported    : 0
max_depth         : 2
cur_path          : 0
pending_favs      : 0
pending_total     : 0
variable_paths    : 18
stability         : 98.40%
bitmap_cvg        : 1.62%
unique_crashes    : 4
unique_hangs      : 0
last_path         : 1529615159
last_crash        : 1529615157
last_hang         : 0
execs_since_crash : 178078
exec_timeout      : 20
afl_banner        : avro-fuzz
afl_version       : 2.52b
target_mode       : default

```

**Comment by laslowh** [comment]
So here's some output from a crash:

**Comment by poros** [comment]
Thank you for the report! Never happened to us so far. We'll try to reproduce.

**Comment by laslowh** [comment]
So apparently there's a movement afoot to add the ability to [handle out of memory conditions](https://github.com/rust-lang/rust/issues/48043). However, as far as I can tell, it's not even a nightly feature yet.  As an interim step, would you consider putting an optional upper bound on the size of any allocations that happen with a "user specified" size?

**Comment by laslowh** [comment]
Ok, so here's a simple program that can reproduce two different types of crashes:

```
extern crate afl;
extern crate avro_rs;
use avro_rs::{from_avro_datum, Schema};

static SCHEMA: &'static str = r#"
            {
                "type": "record",
                "name": "test",
                "fields": [
                    {"name": "a", "type": "long", "default": 42},
                    {"name": "b", "type": "string"}
                ]
            }
        "#;

fn main() {
    let causes_out_of_memory : &[u8] = &[0x3e,0x15,0xff,0x1f,0x15,0xff];
    let causes_left_shift_overflow: &[u8] = &[
        0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1,
        0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1, 0xe1,
        0xe1, 0xe1, 0xe1,
    ];
    let schema = Schema::parse_str(SCHEMA).unwrap();
    let _ = from_avro_datum(&schema, &mut &causes_out_of_memory[..], None);
}
```
Change 'causes_out_of_memory' to 'causes_left_shift_overflow' to see the other one.

**Comment by flavray** [comment]
Thanks for the report!
I've just merged #42 which should fix both issues (at least, the code in the last comment does not panic anymore). Let us know if you find any other crashes when using the library, we've definitely not battle-tested it against bogus input 🙂
